### PR TITLE
Shortcut fixes for #697

### LINKED ISSUE #697: New code editor: several Edit menu options are not working (see note)
The following Edit menu options do not work on Windows/OSX:
cut, copy, paste, select all, undo, redo

Note: the hotkeys work

### PR DESCRIPTION
**Pull request by shaddockh**
This hooks up the editor menu shortcuts properly and fixes #697

Edit->Undo 
Edit->Redo
Edit->Cut
Edit->Copy
Edit->Paste
Edit->Select All
Edit->Format Code

Testing was done on OSX, but this needs to also be tested in Linux and Windows since the shortcut keys are different (CMD becomes CTRL in Windows and Linux).

This also fixes some linter errors in the TS.

### CODE REVIEW

File: Script/AtomicEditor/ui/Shortcuts.ts [hunk 1]
```diff
@@ -38,7 +38,7 @@ class Shortcuts extends Atomic.ScriptObject {
     }
 
     //this should be moved somewhere else...
-    invokePlayOrStopPlayer(debug:boolean = false) {
+    invokePlayOrStopPlayer(debug: boolean = false) {
         this.sendEvent(EditorEvents.SaveAllResources);
```
Comment (JoshEngebretson) [inline Script/AtomicEditor/ui/Shortcuts.ts]: Thanks for addressing some linting

File: Script/AtomicEditor/ui/Shortcuts.ts [hunk 2]
```diff
@@ -164,7 +164,7 @@ class Shortcuts extends Atomic.ScriptObject {
         }
     }
 
-    invokeResourceFrameShortcut(shortcut: string) {
+    invokeResourceFrameShortcut(shortcut: Editor.EditorShortcutType) {
         if (!ToolCore.toolSystem.project) return;
```
Comment (JoshEngebretson) [inline Script/AtomicEditor/ui/Shortcuts.ts]: Strongly typed, ftw

File: Script/AtomicWebViewEditor/clientExtensions/languageExtensions/typescript/TypescriptLanguageExtension.ts [hunk 3]
```diff
@@ -288,6 +288,21 @@ export default class TypescriptLanguageExtension implements Editor.ClientExtensi
...
+     */
+    formatCode() {
+        if (this.active) {
+            const action = this.editor.getAction("editor.action.format");
+            if (action && action.isSupported()) {
+                if (this.editor.getSelection().isEmpty()) {
+                    this.editor.setSelection(this.editor.getModel().getFullModelRange());
+                }
+                action.run();
+            }
```
Comment (JoshEngebretson) [inline Script/AtomicWebViewEditor/clientExtensions/languageExtensions/typescript/TypescriptLanguageExtension.ts]: Fancy footwork here, 👣 , nice one :)

File: Script/AtomicWebViewEditor/editor/editorCommands.ts [hunk 4]
```diff
@@ -56,6 +56,10 @@ export function configure(fileExt: string, filename: string) {
         filename: filename,
         editor: monacoEditor
     });
+    
+    // Override CMD/CTRL+I since that is going to be used for Format Code and in the editor it is assigned to something else
+    monacoEditor.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyCode.KEY_I, null, null);
+
```
Comment (JoshEngebretson) [inline Script/AtomicWebViewEditor/editor/editorCommands.ts]: Good commenting

File: Script/TypeScript/EditorWork.d.ts [hunk 5]
```diff
@@ -514,3 +515,10 @@ declare module Editor.ClientExtensions {
         addCustomHostRoutine(routineName: string, callback: (...any) => void);
     }
 }
+
+declare module Editor {
+    /**
+     * Valid editor shortcuts that can be called from menus
+     */
+    export type EditorShortcutType = "cut" | "copy" | "paste" | "undo" | "redo" | "close" | "frameselected" | "selectall";
+}
```
Comment (JoshEngebretson) [inline Script/TypeScript/EditorWork.d.ts]: I'd really like to spend some more time in TS, very nice

File: Source/AtomicWebView/WebClient.cpp [hunk 6]
```diff
@@ -353,6 +353,8 @@ class WebClientPrivate : public CefClient,
         browserSettings.file_access_from_file_urls = STATE_ENABLED;
         browserSettings.universal_access_from_file_urls = STATE_ENABLED;
         browserSettings.web_security = WebBrowserHost::GetWebSecurity() ? STATE_ENABLED : STATE_DISABLED;
+        browserSettings.javascript_access_clipboard = STATE_ENABLED;
+        browserSettings.javascript_dom_paste = STATE_ENABLED;
```
Comment (JoshEngebretson) [inline Source/AtomicWebView/WebClient.cpp]: Look what treasure ye can find when digging ☠️

File: Source/AtomicEditor/Editors/JSResourceEditor.cpp [hunk 7]
```diff
@@ -177,6 +177,39 @@ bool JSResourceEditor::OnEvent(const TBWidgetEvent &ev)
         if (ev.ref_id == TBIDC("close"))
         {
             RequestClose();
+        } else if (ev.ref_id == TBIDC("undo")) {
+            // Need to physically send the CTRL/CMD+Z to the browser so that
```
Comment (JoshEngebretson) [inline Source/AtomicEditor/Editors/JSResourceEditor.cpp]: Crunchy block of code, happy that can be baked down here without invading a bunch of source files 👍

**Comment by JimMarlowe** [comment]
Compiles fine on linux, it does not fix linux issue with the menu shortcuts (nor did it claim it had).

**Comment by JoshEngebretson** [comment]
I tested on Windows and Linux, everything works other than Edit->Undo,  Edit->Redo on Windows

**Comment by JoshEngebretson** [comment]
@JimMarlowe I tested on Linux and it does fix the issues here, sure you were on the branch?

**Comment by JoshEngebretson** [comment]
Oh, I think you mean the accelerators, I only asked about branch because I had just "tested" on the wrong one :)  

On Linux, invoking the main frame Edit menu "Cut, Copy, Paste" is working here, I didn't test accelerators.  Though, reminds me that issue should be bumped to Build 1 to at least look into for it.

**Comment by JimMarlowe** [comment]
I believe I am, I spot checked some of the files and the look the same.

**Comment by shaddockh** [comment]
for edit redo and undo on the mac,  I needed to spend some time in the chrome debug tools to ensure that the key events were actually getting passed in correctly.  This probably needs to be done in windows too.  Do the keyboard shortcuts work in windows?  Or are both the menus and key accelerators broken?

I believe I should be able to get my windows environment working this week if necessary to try to debug.

**Comment by JoshEngebretson** [comment]
CEF doesn't have a keyboard abstraction which means you are left fiddling/mapping/simulating "native key events".  It doesn't surprise me that Undo/Redo might need special handling on a platform.

The issue addressed by this PR is 100% working here, minus Undo/Redo from main frame menu items on Windows.  #1154 is a separate issue, solely affecting Linux accelerator keys.

**Comment by JoshEngebretson** [comment]
@shaddockh Ok,  here's the super complicated instructions for getting running under VS2015 ;)

https://github.com/AtomicGameEngine/AtomicGameEngine/wiki/Building-the-Atomic-Editor-using-Visual-Studio-2015

Going to need to spend a few hours getting build setup for VS2017 now that it is in RC and from all indications is a significant improvement over 2015 (including the install process).  Filed as #1188

**Comment by CTrauma** [comment]
Building on a few distros and notice that the shortcut keys all seem to work but the menu items for undo/redo have no effect

**Comment by JoshEngebretson** [comment]
@CTrauma thanks, I rechecked, **Undo/Redo** on **Windows/Linux** from **edit menu** not working here, I'll update #1188 with the keyboard testing since that is getting confusing here

**Comment by JoshEngebretson** [comment]
Does that last commit have anything to do with the Undo/Redo issue, or something else?

**Comment by shaddockh** [comment]
So, the exact same key events were getting registered by the web view either by sending the key events through the menu item or pressing the shortcut keys, but the menu item one wasn't working.  It appears that by sending a ```focus``` event, this has been fixed under windows.  Can someone else try this out under Windows as well as Linux?  I'm calling the ```SendFocusEvent``` in both of these platforms now.

File: Source/AtomicEditor/Editors/JSResourceEditor.cpp [hunk 8]
```diff
@@ -202,6 +203,7 @@ bool JSResourceEditor::OnEvent(const TBWidgetEvent &ev)
             map["ForceSuperDown"] = true;
             #else
             map[KeyUp::P_QUALIFIERS] = QUAL_SHIFT | QUAL_CTRL;
+			webClient_->SendFocusEvent();
             #endif
```
Comment (JoshEngebretson) [inline Source/AtomicEditor/Editors/JSResourceEditor.cpp]: White spaceeeeeeee 🚀

Guessing VS code editor configured for tabs, that has bitten me so many times

tabsspaces

Comment (shaddockh) [inline Source/AtomicEditor/Editors/JSResourceEditor.cpp]: bleh... I didn't get VS2015 set up correctly with whitespace... will fix.

Comment (shaddockh) [inline Source/AtomicEditor/Editors/JSResourceEditor.cpp]: Also didn't help that the diff view in SourceTree was converting tabs to spaces so it didn't show up... both have now been changed.

**Comment by JoshEngebretson** [comment]
I wrote a cheesy ```build:lint``` task which scans the C++/C# code and is capable of fixing tab issues if "configured", which involves changing this line: 

https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Build/Scripts/BuildLint.js#L7

Testing Windows/Linux, will land if they check out.. thanks!

**Comment by shaddockh** [comment]
Sorry, just saw the comment on the commit prior to this last one... no, it wasn't the undo/redo issue, but it was a fine tuning of the format code menu shortcut.

**Comment by JoshEngebretson** [comment]
Glorious victory, undo/redo working from menu here ⚔ 

One issue I ran into on Linux only, Ctrl-I to format code causes the below error, while the menu item works great, this could be a follow up issue if it isn't obvious

Linux:
```
[Tue Nov 22 20:16:54 2016] INFO: WebViewJS: TypeError: Cannot read property 'apply' of null (atomic:///home/jenge/Dev/atomic/PullRequests/AtomicGameEngine/Artifacts/AtomicEditor/Resources/ToolData/CodeEditor/source/monaco/vs/editor/editor.main.js:66)
[1122/201654:INFO:CONSOLE(66)] "TypeError: Cannot read property 'apply' of null", source: atomic:///home/jenge/Dev/atomic/PullRequests/AtomicGameEngine/Artifacts/AtomicEditor/Resources/ToolData/CodeEditor/source/monaco/vs/editor/editor.main.js (66)
```

**Comment by shaddockh** [comment]
yeah, it looks like that was happening on other platforms too.  This should fix it.

**Comment by JoshEngebretson** [comment]
Awesome, will land when green, thanks! ✔️

**Pull request merged**